src/Microsoft.Extensions.FileProviders.Physical/PhysicalFilesWatcher.cs

```diff
@@ -33,30 +35,88 @@ public PhysicalFilesWatcher(string root, FileSystemWatcher fileSystemWatcher)
+            filter = NormalizePath(filter);
+
+            IChangeToken changeToken;
+            var isWildCard = filter.IndexOf('*') != -1;
+            if (isWildCard || IsDirectoryPath(filter))
+            {
+                changeToken = ResolveFileTokensForGlobbingPattern(filter);
+            }
+            else
+            {
+                changeToken = GetOrAddChangeToken(filter);
+            }
 
-            FileChangeToken changeToken;
-            if (!_tokenCache.TryGetValue(pattern, out changeToken))
+            lock (_lockObject)
             {
-                changeToken = _tokenCache.GetOrAdd(pattern, new FileChangeToken(pattern));
-                lock (_lockObject)
+                if (_matchInfoCache.Count > 0 && !_fileWatcher.EnableRaisingEvents)
                 {
-                    if (_tokenCache.Count > 0 && !_fileWatcher.EnableRaisingEvents)
-                    {
-                        // Perf: Turn on the file monitoring if there is something to monitor.
-                        _fileWatcher.EnableRaisingEvents = true;
-                    }
+                    // Perf: Turn on the file monitoring if there is something to monitor.
+                    _fileWatcher.EnableRaisingEvents = true;
                 }
             }
 
             return changeToken;
         }
 
+        private IChangeToken GetOrAddChangeToken(string filePath)
+        {
+            ChangeTokenInfo tokenInfo;
+            if (!_matchInfoCache.TryGetValue(filePath, out tokenInfo))
+            {
+                var cancellationTokenSource = new CancellationTokenSource();
+                var cancellationChangeToken = new CancellationChangeToken(cancellationTokenSource.Token);
+                tokenInfo = new ChangeTokenInfo(cancellationTokenSource, cancellationChangeToken);
+                tokenInfo = _matchInfoCache.GetOrAdd(filePath, tokenInfo);
+            }
+
+
+            IChangeToken changeToken = tokenInfo.ChangeToken;
+            if (_pollForChanges)
+            {
+                // The expiry of CancellationChangeToken is controlled by this type and consequently we can cache it.
+                // PollingFileChangeToken on the other hand manages its own lifetime and consequently we cannot cache it.
+                changeToken = new CompositeFileChangeToken(
+                    new[]
+                    {
+                        changeToken,
+                        new PollingFileChangeToken(new FileInfo(filePath))
+                    });
+            }
+
+            return changeToken;
+        }
+
+        private IChangeToken ResolveFileTokensForGlobbingPattern(string filter)
+        {
+            var matcher = new Matcher(StringComparison.OrdinalIgnoreCase);
+            matcher.AddInclude(filter);
+
+            var directoryBase = new DirectoryInfoWrapper(new DirectoryInfo(_root));
+            var result = matcher.Execute(directoryBase);
+
+            var changeTokens = new List<IChangeToken>();
```

Comment:
Minor perf optimization. You can initialize the list to the result.Files.Count capacity so it doesn't have to resize at all